### PR TITLE
feat: Replicated Growable Array CRDT and Tests

### LINKED ISSUE #83: Implement built-in CRDTs: RGA (array type)
### Description
Same as the description of #82. This issue tracks the implementation and testing of RGA.

### Task
- implement
- test
- document how this type works and its usage

### Resources
- https://www.sciencedirect.com/science/article/abs/pii/S0743731510002716 original paper
- https://github.com/Horusiath/crdt-examples/blob/master/Crdt/commutative/Rga.fs Bartosz's implementation
- https://josephg.com/blog/crdts-go-brrr/ this seminal blogpost in the CRDT circle also covers how RGA works

### PR DESCRIPTION
**Pull request by JanLewDev**
Resolves #83

### CODE REVIEW

File: packages/crdt/src/builtins/RGA/index.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    isTombstone(element: RGAElement<T>): boolean {
+        return element.value === null;
+    }
+
+    nextSeqNr(sequencer: Identifier): Identifier {
+        return { counter: sequencer.counter + 1, nodeId: sequencer.nodeId };
+    }
+
+    // Function to map a logical index (ignoring tombstones) to a physical index in the elements array
+    indexWithTombstones(index: number): number {
```
Comment (guiltygyoza) [inline packages/crdt/src/builtins/RGA/index.ts:59]: this function maps virtual pointer (virtual id) to its actual index right? may be good practice to name it as such

File: packages/crdt/src/builtins/RGA/index.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            if (!this.isTombstone(this._elements[offset])) index--;
+            offset++;
+        }
+        if (index > 0) {
+            throw new RangeError("Index not found");
+        }
+        return offset;
+    }
+
+    read(index: number): T | null {
```
Comment (guiltygyoza) [inline packages/crdt/src/builtins/RGA/index.ts:73]: this function reads the value at an (actual) index. it seems that a function for reading the entire array out (no tombstoned values; sort by RGA's rules) is missing?

File: packages/crdt/src/builtins/RGA/index.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        this._elements = [
+            new RGAElement<T>({ counter: 0, nodeId: "" }, null),
+        ]
+    }
+
+    isTombstone(element: RGAElement<T>): boolean {
+        return element.value === null;
+    }
+
+    nextSeqNr(sequencer: Identifier): Identifier {
```
Comment (guiltygyoza) [inline packages/crdt/src/builtins/RGA/index.ts:54]: some functions here should be private methods. I'll defer to @d-roak for coding style guideline

File: packages/crdt/src/builtins/RGA/index.ts (hunk 2 above)
Comment (JanLewDev) [reply packages/crdt/src/builtins/RGA/index.ts:77]: I think getArray() does that now

Comment (guiltygyoza) [reply packages/crdt/src/builtins/RGA/index.ts:77]: gotcha!

File: packages/crdt/src/builtins/RGA/index.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        while (this.isTombstone(this._elements[i])) i++;
+        return this._elements[i].value;
+    }
+
+    // Function to find the physical index of an element given the virtual id
+    private indexOfVId(ptr: Identifier): number {
+        for (let offset = 0; offset < this._elements.length; offset++) {
+            if (
+                ptr.counter === this._elements[offset].vid.counter &&
+                ptr.nodeId === this._elements[offset].vid.nodeId
```
Comment (d-roak) [inline packages/crdt/src/builtins/RGA/index.ts:94]:
```suggestion
        for (let offset = 0; offset < this.elements.length; offset++) {
            if (
                ptr.counter === this.elements[offset].vid.counter &&
                ptr.nodeId === this.elements[offset].vid.nodeId
```

File: packages/crdt/src/builtins/RGA/index.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            return;
+        }
+        this._elements.splice(insertIdx, 0, element);
+    }
+
+    // Function to delete an element from the RGA
+    delete(index: number): void {
+        let i = this.indexWithTombstones(index);
+        while (this.isTombstone(this._elements[i])) i++;
+        this._elements[i].value = null;
```
Comment (d-roak) [inline packages/crdt/src/builtins/RGA/index.ts:139]:
```suggestion
        while (this.isTombstone(this.elements[i])) i++;
        this.elements[i].value = null;
```

File: packages/crdt/src/builtins/RGA/index.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    // Function to update the value of an element
+    update(index: number, value: T): void {
+        let i = this.indexWithTombstones(index);
+        while (this.isTombstone(this._elements[i])) i++;
+        this._elements[i].value = value;
```
Comment (d-roak) [inline packages/crdt/src/builtins/RGA/index.ts:146]:
```suggestion
        while (this.isTombstone(this.elements[i])) i++;
        this.elements[i].value = value;
```

File: packages/crdt/src/builtins/RGA/index.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        for (let i = 1; i < peerRGA.elements().length; i++) {
+            this.insertElement(peerRGA.elements()[i]);
+        }
+
+        this._sequencer = {
+            counter: Math.max(
+                this._sequencer.counter,
+                peerRGA._sequencer.counter
+            ),
+            nodeId: this._sequencer.nodeId,
```
Comment (d-roak) [inline packages/crdt/src/builtins/RGA/index.ts:160]:
```suggestion
        this.sequencer = {
            counter: Math.max(
                this.sequencer.counter,
                peerRGA.sequencer.counter
            ),
            nodeId: this.sequencer.nodeId,
```

**Review by d-roak — CHANGES_REQUESTED** [review]

**Review by d-roak — APPROVED** [review]
lgtm

**Pull request merged**